### PR TITLE
Disable Optimizing try finally when we bailout on early exit

### PR DESCRIPTION
**Pull request by meg-gupta**

### CODE REVIEW

File: lib/Backend/JITTimeProfileInfo.h [hunk 1]
```diff
@@ -63,6 +63,7 @@ class JITTimeProfileInfo
     bool IsLoopImplicitCallInfoDisabled() const;
     bool IsPowIntIntTypeSpecDisabled() const;
     bool IsTagCheckDisabled() const;
+    bool IsOptimizeTryFinally() const;
```
Comment (rajatd) [inline lib/Backend/JITTimeProfileInfo.h]: >IsOptimizeTryFinally [](start = 9, length = 20)

Can you call this function IsOptimizeTryFinallyDisabled?

Comment (meg-gupta) [inline lib/Backend/JITTimeProfileInfo.h]: Thanks! Done.

**Comment by rajatd** [comment]
:shipit:

File: lib/Backend/BailOut.cpp [hunk 2]
```diff
@@ -2200,6 +2200,12 @@ void BailOutRecord::ScheduleFunctionCodeGen(Js::ScriptFunction * function, Js::S
                     rejitReason = RejitReason::PowIntIntTypeSpecDisabled;
                 }
             }
+            case IR::BailOutOnEarlyExit:
+            {
+                profileInfo->DisableOptimizeTryFinally();
```
Comment (LouisLaf) [inline lib/Backend/BailOut.cpp]: You should follow the same pattern as above:
if (already disabled)
   reThunk to the latest JIT'd code
else
   disable it

File: lib/Backend/BailOut.cpp [hunk 3]
```diff
@@ -2200,6 +2200,12 @@ void BailOutRecord::ScheduleFunctionCodeGen(Js::ScriptFunction * function, Js::S
...
+                rejitReason = RejitReason::OptimizeTryFinallyDisabled;
+                break;
+            }
```
Comment (LouisLaf) [inline lib/Backend/BailOut.cpp]: Add the same code to the JIT loop body version of this function.

Comment (rajatd) [inline lib/Backend/BailOut.cpp]: we don't optimize try finally's in jitted loop bodies

File: lib/Backend/BailOut.cpp [hunk 4]
```diff
@@ -2199,6 +2199,20 @@ void BailOutRecord::ScheduleFunctionCodeGen(Js::ScriptFunction * function, Js::S
                     profileInfo->DisablePowIntIntTypeSpec();
                     rejitReason = RejitReason::PowIntIntTypeSpecDisabled;
                 }
+                break;
```
Comment (LouisLaf) [inline lib/Backend/BailOut.cpp]: >break [](start = 16, length = 5)

Yikes!  Good catch....

**Comment by LouisLaf** [comment]
:shipit:

**Pull request merged**